### PR TITLE
🔖 feat: enforce required tag categories on Save Changes in metadata drawer

### LINKED ISSUE #2535: Enforce required tag categories on Save Changes button in metadata drawer
> *This was generated by AI during triage.*

## Summary

Tag categories support an `isRequired` flag, and the `fix/required-tag-ux` branch already surfaces this with an asterisk on the field. However, a user can still click **Save Changes** with no options selected for a required category — the flag is purely cosmetic today. This ticket enforces the constraint on the frontend Save Changes button.

## Motivation

A required tag category means every collection item must carry at least one selection from it before saving. Without enforcement, the asterisk is misleading: it signals a requirement but never actually blocks an invalid save. The publish flow already warns about this (non-blocking, `PublishButton.tsx` lines 79–83) — this brings the save flow to parity.

## Acceptance Criteria

- [ ] Save Changes is **disabled** in `MetadataEditorStateDrawer` when any required tag category has zero options selected
- [ ] This applies immediately on drawer open (existing items with unfilled required categories are blocked, not just new interactions)
- [ ] Each unfilled required category shows an inline error message: **"At least one option must be selected"** on the `MultiSelect` field
- [ ] Non-required categories and pages without tag categories are unaffected
- [ ] No changes to tRPC backend validation
- [ ] `CollectionEditorStateDrawer` is not in scope (it manages filter definitions, not tag selection)

### New hook — `useRequiredTagsValidation`

Create a shared hook (suggested location: `apps/studio/src/features/editing-experience/hooks/useRequiredTagsValidation.ts`):

```ts
useRequiredTagsValidation({ resourceId, siteId, tagged })
// → { unfilledRequiredCategories: TagCategory[], isValid: boolean }
```

- Fetches categories via `trpc.collection.getCollectionTags.useQuery({ resourceId, siteId })`
- Mirrors the logic already in `PublishButton.tsx` lines 79–83
- Keep it free of drawer-specific concerns so it can be reused elsewhere

### `MetadataEditorStateDrawer.tsx`
File: `apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx`

- Call `useRequiredTagsValidation` with `resourceId` (pageId) and `tagged` from `previewPageState.page`
- Pass `isValid` as an extra `isDisabled` condition into `SaveButton` (alongside the existing `!isEmpty(errors)` from `useBuilderErrors`)
- Only call the hook when the layout is CollectionPage / CollectionLink; treat as valid otherwise

### `JsonFormsTaggedControl.tsx`
File: `apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsTaggedControl.tsx`

- For each tag category rendered, check `tagIsRequired && currentTagCategoryOptions.length === 0`
- Set `<FormControl isInvalid={...}>` and add `<FormErrorMessage>At least one option must be selected</FormErrorMessage>`
- This is purely presentational — it reads from the already-fetched `tags` data; no new queries

### Why not inject into `ErrorProvider`?

`FormBuilder.tsx` calls `setErrors(groupBy(errors, "instancePath"))` on every `onChange`, which is a full replacement. Injecting custom required-tag errors would be wiped on the next field change. The separate hook + extra `isDisabled` prop keeps `ErrorProvider` untouched and is consistent with how the existing publish-flow validation works.

### Reference

- Existing validation pattern: `PublishButton.tsx` lines 79–83
- `ErrorProvider`: `apps/studio/src/features/editing-experience/components/form-builder/ErrorProvider.tsx`
- `isRequired` schema definition: `packages/components/src/types/page.ts` lines 33–77

### PR DESCRIPTION
**Pull request by adriangohjw**
Closes #2535

## Summary

- Adds `useRequiredTagsValidation` hook that fetches tag categories and returns `{ isValid, unfilledRequiredCategories }` — mirrors the validation pattern already in the publish flow
- `MetadataEditorStateDrawer`: Save Changes is disabled when any required tag category has zero selections. Uses a `TagsAwareSaveButton` component that only mounts (and only fires the tRPC query) for article/link layout pages — no overhead for other page types
- `JsonFormsTaggedControl`: shows an inline `"At least one option must be selected"` error message under any required tag category that has no selection

## Test plan

- [ ] Open a collection item where a required tag category has no options selected → Save Changes button is disabled
- [ ] Select at least one option in each required category → Save Changes becomes enabled
- [ ] Inline error message appears under unfilled required categories and disappears on selection
- [ ] Pages without tag categories (content, index, homepage, etc.) are unaffected — no tRPC query fires
- [ ] Non-required tag categories do not block saving or show an error

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Greptile Summary</h3></summary>

This PR enforces required tag category validation on the Save action for collection pages and links. Previously, save was gated only on form errors and a valid `ref`; now it is also gated on every required tag category having at least one selection, with the validation centralized in a new `validateRequiredTags` utility and a `useCollectionTags` hook.

- Adds `validateRequiredTags` (with unit tests) and `useCollectionTags` / `useSuspenseCollectionTags` as a single source of truth for the tRPC query; both `LinkEditorDrawer` and `MetadataEditorStateDrawer` now disable their Save button during query loading and whenever any required tag category is unfilled.
- `JsonFormsTaggedControl` gains an inline `\"At least one option must be selected\"` error message under each unfilled required category, consistent with the save-button gate.
- E2E tests cover the enabled/disabled button state and assert the selected tags are persisted to the blob for both the link and article page drawers.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — validation logic is correct, loading-state races are guarded, and the change is well-covered by unit and E2E tests.

All four issues flagged by the previous review round have been addressed: empty-option categories no longer block save, the tRPC query is properly gated to collection items only, and both drawers correctly disable the save button during the initial query flight. The remaining observations are minor API-design and test-hygiene points that do not affect runtime correctness.

No files require special attention; the minor interface mismatch in `useCollectionTags.ts` and the orphaned-row concern in `collection.ts` are worth a follow-up but do not block the feature.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts | New hook that centralises tRPC collection tag fetching; `useSuspenseCollectionTags` accepts `UseCollectionTagsInput` (which includes `enabled`) but silently ignores that field. |
| apps/studio/src/features/editing-experience/utils/validateRequiredTags.ts | Pure validation utility; correctly skips categories with zero options (preventing a permanent block), and returns `isValid` / `unfilledRequiredCategories`. |
| apps/studio/src/features/editing-experience/components/Drawer/LinkEditorDrawer.tsx | Adds tag validation to the Save button; correctly gates on `isCollectionTagsLoading` to prevent bypassing validation during query flight. |
| apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx | Adds `isCollectionItem` gating and a `TagsAwareSaveButton` / `SaveButton` split; loading state is correctly locked via `isTagsValid={!isCollectionItem || !isCollectionTagsLoading}`. |
| apps/studio/tests/e2e/fixtures/collection.ts | E2E fixtures for collection setup; `deleteCollection` cascades over Resource rows but leaves orphaned Blob and Version records in the test database. |
| apps/studio/src/features/editing-experience/utils/__tests__/validateRequiredTags.test.ts | Comprehensive unit tests covering all edge cases, including required categories with no options and missing `isRequired` flag. |
| apps/studio/tests/e2e/collection/required-tags-validation.test.ts | Thorough E2E tests for both link and article drawers covering the enabled/disabled button states and the persistence of tag selections. |
| apps/studio/tests/e2e/fixtures/collection.po.ts | Clean page-object model for the tag-selection UI with a clear explanation for why `getByLabel` is not used. |
| apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsTaggedControl.tsx | Adds inline `FormErrorMessage` for unfilled required tag categories; `isInvalid` only set for categories that have options, consistent with `validateRequiredTags`. |
| apps/studio/src/features/editing-experience/components/preview/EditLinkPreview.tsx | Trivial refactor to use the new `useSuspenseCollectionTags` hook; no logic change. |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User opens metadata/link drawer] --> B{isCollectionItem?}
    B -- No --> C[SaveButton isTagsValid=true]
    B -- Yes --> D[useCollectionTags fires\nwith enabled:true]
    D --> E{Query loading?}
    E -- Yes --> F[SaveButton isTagsValid=false\nbutton disabled]
    E -- No --> G{collectionTags.length > 0?}
    G -- No --> H[SaveButton isTagsValid=true\nno categories to validate]
    G -- Yes --> I[TagsAwareSaveButton\ncalls validateRequiredTags]
    I --> J{All required categories\nhave ≥1 selection?}
    J -- Yes --> K[SaveButton isTagsValid=true\nbutton enabled]
    J -- No --> L[SaveButton isTagsValid=false\nbutton disabled]
    L --> M[JsonFormsTaggedControl\nshows inline error per unfilled category]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[User opens metadata/link drawer] --> B{isCollectionItem?}
    B -- No --> C[SaveButton isTagsValid=true]
    B -- Yes --> D[useCollectionTags fires\nwith enabled:true]
    D --> E{Query loading?}
    E -- Yes --> F[SaveButton isTagsValid=false\nbutton disabled]
    E -- No --> G{collectionTags.length > 0?}
    G -- No --> H[SaveButton isTagsValid=true\nno categories to validate]
    G -- Yes --> I[TagsAwareSaveButton\ncalls validateRequiredTags]
    I --> J{All required categories\nhave ≥1 selection?}
    J -- Yes --> K[SaveButton isTagsValid=true\nbutton enabled]
    J -- No --> L[SaveButton isTagsValid=false\nbutton disabled]
    L --> M[JsonFormsTaggedControl\nshows inline error per unfilled category]
```

</a>
</details>

<sub>Reviews (11): Last reviewed commit: ["Merge branch &#39;collection-tags-master&#39; in..."](https://github.com/opengovsg/isomer/commit/c9fc5b12ce3fe7baf7e8da5f9e673513a4823b7b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39746598)</sub>

### CODE REVIEW

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx [hunk 1]
```diff
@@ -90,6 +91,10 @@ export default function MetadataEditorStateDrawer(): JSX.Element {
 
   const validateFn = ajv.compile<Static<typeof metadataSchema>>(filteredSchema)
 
+  const isTaggableLayout =
+    previewPageState.layout === ISOMER_USABLE_PAGE_LAYOUTS.Article ||
+    previewPageState.layout === ISOMER_USABLE_PAGE_LAYOUTS.Link
+
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: issue: i don't think this is correct - we define tags on the collection index, so it's actually the presence of a parent collection that makes them a taggable layout

Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: nice catch - i refactor to use `apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts` as the single source of truth around the codebase

**Comment by seaerchin** [comment]
@greptileai

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx [hunk 2]
```diff
@@ -44,6 +47,12 @@ export default function MetadataEditorStateDrawer(): JSX.Element {
   } = useEditorDrawerContext()
 
   const { pageId, siteId } = useQueryParse(pageSchema)
+
+  const { data: collectionTags = [] } = useCollectionTags({
+    resourceId: pageId,
+    siteId,
+  })
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: note: good catch, i actually reverted this to make it more readable lol but okay to add it back

**Comment by mvincentong** [comment]
I think the required-tag save guard still misses collection links. This wires validation into `MetadataEditorStateDrawer`, but link items are edited through `LinkEditorDrawer` on `/sites/[siteId]/links/[linkId]`; that Save button is still disabled only by form errors / missing `ref`, then the mutation submits `...link` directly. Since link items also have `tagged` and the PR summary calls out article/link layout pages, a link with an unfilled required tag category can still be saved. Could we reuse `validateRequiredTags` / `useCollectionTags` in `LinkEditorDrawer` too, or move the validation into a shared save path for both collection page and link items?

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx [hunk 3]
```diff
@@ -44,6 +47,17 @@ export default function MetadataEditorStateDrawer(): JSX.Element {
...
   const { pageId, siteId } = useQueryParse(pageSchema)
+
+  const isCollectionItem =
+    previewPageState.layout === ISOMER_USABLE_PAGE_LAYOUTS.Article ||
+    previewPageState.layout === ISOMER_USABLE_PAGE_LAYOUTS.Link
+
+  const { data: collectionTags = [] } = useCollectionTags({
+    resourceId: pageId,
+    siteId,
+    enabled: isCollectionItem,
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: comment: this doesn't seem correct still - we fire when layout is either Article (can belong at root, inside folder or inside collection) or link (only inside collection)

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsCategoryIdControl.tsx [hunk 4]
```diff
@@ -68,7 +68,7 @@ function SuspendableJsonFormsCategoryIdSelect({
 
 export function JsonFormsCategoryIdControl({
   description,
-  required,
+  required: _required,
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsCategoryIdControl.tsx]: comment: can remove if not used

File: apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  siteId,
+  enabled = true,
+}: UseCollectionTagsInput) {
+  return trpc.collection.getCollectionTags.useQuery(
+    { resourceId, siteId },
+    { enabled },
+  )
+}
+
+export function useCollectionTagsSuspense({
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts]: comment: prefer `useSuspenseCollectionTags` but either way is fine

**Review by seaerchin — COMMENTED** [review]
1 blocking issue but happy to approve after

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx (hunk 3 above)
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: eh wtf i tot i changed this, i think the AI changed back haha my bad sir for reviewing this shit

File: apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts (hunk 5 above)
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/hooks/useCollectionTags.ts]: good call, i just update https://github.com/opengovsg/isomer/pull/2536/commits/cf54a07ada71fff0173b2c5d9f2a3c17296b541e

File: apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsCategoryIdControl.tsx (hunk 4 above)
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsCategoryIdControl.tsx]: donez https://github.com/opengovsg/isomer/pull/2536/commits/cf54a07ada71fff0173b2c5d9f2a3c17296b541e

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx (hunk 3 above)
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx]: updated this to use `type` instead to check for CollectionPage and CollectionLink 
https://github.com/opengovsg/isomer/pull/2536/commits/cf54a07ada71fff0173b2c5d9f2a3c17296b541e

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx [hunk 6]
```diff
@@ -181,7 +197,20 @@ export default function MetadataEditorStateDrawer(): JSX.Element {
...
+                isLoading={isPending}
+                onClick={handleSaveChanges}
+                tags={collectionTags}
+                tagged={(previewPageState.page as { tagged?: string[] }).tagged}
+              />
+            ) : (
+              <SaveButton
+                isLoading={isPending}
+                onClick={handleSaveChanges}
+                isTagsValid={!isCollectionItem || !isCollectionTagsLoading}
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:211]: nit: can we just reuse `ButtonProps` and do a `Pick<isLoading | onClick | isDisabled>`?

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx [hunk 7]
```diff
@@ -192,20 +221,40 @@ export default function MetadataEditorStateDrawer(): JSX.Element {
...
+  isLoading,
+  tags,
+  tagged,
+}: {
+  onClick: () => void
+  isLoading: boolean
+  tags: CollectionTags
+  tagged: string[] | undefined
+}) => {
+  const { isValid } = validateRequiredTags(tags, tagged)
```
Comment (seaerchin) [inline apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: comment: i noticed that this wasn't on the backend but i'm assuming that is fine because we are migrating away from category -> tags

**Review by seaerchin — APPROVED** [review]
lgtm

File: apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx (hunk 7 above)
Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: WDYM ah? like we aren't validating again that its valid on the backend again?

Comment (seaerchin) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: idk why it missed the comment but my assumption is that we are migrating away from categories -> tags so this might not be relevant **unless** we now say that we allow setting of required tags

Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: oh this is still required because tags have a new attribute `isRequired` so we would need to ensure that if the tags are required, the user select at least one thing

Comment (seaerchin) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: oh ok! should we re-validate on backend? i scanned and i saw this was a frontend only thing

Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: ahh yeah makes sense to add it - will add it in a following PR

Comment (adriangohjw) [reply apps/studio/src/features/editing-experience/components/Drawer/MetadataEditorStateDrawer.tsx:255]: recorded here https://github.com/opengovsg/isomer/issues/2736
